crates/bevy_reflect/bevy_reflect_derive/src/lib.rs

```diff
@@ -55,6 +58,41 @@ pub fn derive_reflect(input: TokenStream) -> TokenStream {
     }
 }
 
+#[proc_macro_derive(TypeName, attributes(module, type_name))]
+pub fn derive_type_name(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let derive_data = match ReflectDerive::from_input(&ast) {
+        Ok(data) => data,
+        Err(err) => return err.into_compile_error().into(),
+    };
+
+    let s = match derive_data {
```

Comment:
Nit: I'm not sure clippy will like the `s`. Maybe just use `.into()` on the `match`?